A59-audit-logging.md

```diff
@@ -0,0 +1,349 @@
+        "audit_condition": {
+          "description": "Condition for the audit logging to happen."
+            "Same as NONE if ommited.",
+          "type": "string",
+          "enum": ["NONE", "ON_DENY", "ON_ALLOW", "ON_DENY_AND_ALLOW"]
+        },
+        "audit_logger": {
+          "description": "Audit logger configurations",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/audit_logger"
+          }
+        }
+      }
+    },
+  },
+  "required": ["name", "allow_rules"]
+}
+```
+
+For readers that are more familiar with proto messages, the change translates
+to:
+
+```
+message AuthorizationPolicy {
+  …
+
+  message AuditLoggingOptions {
+    enum AuditCondition {
+      NONE = 0;
+      ON_DENY = 1;
+      ON_ALLOW = 2;
+      ON_DENY_AND_ALLOW = 3;
+    }
+
+    AuditCondition audit_condition = 1;
+
+    message AuditLogger {
+      string name = 1;
+      google.protobuf.Struct config = 2;  
+    }
+    
+    repeated AuditLogger audit_loggers = 2;
+  }
+
+  AuditLoggingOptions audit_logging_options = 4;
+}
+```
+
+Note that the definition above is only for illustration purposes and does not
+actually exist anywhere as a `.proto` file. Nor does gRPC process the policy
+as protobuf by any means.
+
+The authorization policy is backed by two RBAC filters, so the audit logger
+configurations will be duplicated in both generated RBAC filters. How the
+audit condition gets translated into two conditions in those RBACs is less
+straightforward and so is explained below.
+
+First of all, note that the RBAC with `DENY` is placed before the RBAC with
+`ALLOW`. We assume users will want to audit one particular RPC exactly once
+if the condition is met.
+
+If users want to audit on deny, then both of the RBACs will have `ON_DENY` as
+the audit condition. The `ALLOW` RBAC will not be evaluated so audit logging
+will never happen twice. 
+
+If users want to audit on allow, then the `DENY` RBAC will have no audit enabled
+and the `ALLOW` RBAC with `ON_ALLOW`. So audit logging will at most happen once
+when the RPC passes through both RBACs.
+
+If users want to audit on both cases, then the `DENY` RBAC needs `ON_DENY` and
+the `ALLOW` RBAC needs `ON_DENY_AND_ALLOW`.
+
+Following is the table summarizing the combinations.
+
+|Authorization Policy  |DENY RBAC          |ALLOW RBAC           |
+|----------------------|-------------------|---------------------|
+|NONE                  |NONE               |NONE                 |
+|ON_DENY               |ON_DENY            |ON_DENY              |
+|ON_ALLOW              |NONE               |ON_ALLOW             |
+|ON_DENY_AND_ALLOW     |ON_DENY            |ON_DENY_AND_ALLOW    |
+
+Again this is a subset of what users can technically do with xDS RBACs in the way
+described in the earlier section. But we think this represents the most common
+use cases.
+
+### Built-in logger types
+
+We plan to implement the stdout logger as a built-in logger type. More types of
```

Comment:
What is "the stdout logger"? It needs to be defined. What is the type used in the Struct? What output format does it use? Does it have any configuration (e.g., to use stderr)?